mono/mini/interp/transform.c

```diff
@@ -7556,392 +7563,393 @@ interp_local_deadce (TransformData *td, int *local_ref_count)
+
+		INTERP_FOLD_BINOP_FULL (MINT_REM_I4, LOCAL_VALUE_I4, i, %, gint32, val2->i != 0 && (val1->i != G_MININT32 || val2->i != -1));
+		INTERP_FOLD_BINOP_FULL (MINT_REM_I8, LOCAL_VALUE_I8, l, %, gint64, val2->l != 0 && (val1->l != G_MININT64 || val2->l != -1));
+		INTERP_FOLD_BINOP_FULL (MINT_REM_UN_I4, LOCAL_VALUE_I4, i, %, guint32, val2->i != 0);
+		INTERP_FOLD_BINOP_FULL (MINT_REM_UN_I8, LOCAL_VALUE_I8, l, %, guint64, val2->l != 0);
 
 		default:
-			goto cfold_failed;
+			return ins;
 	}
 
-	// We were able to compute the result of the ins instruction. We store the
-	// current value for the top of the stack and, if possible, try to replace the
-	// instructions that are part of this unary operation with a single LDC.
+	// We were able to compute the result of the ins instruction. We replace the binop
+	// with a LDC of the constant. We leave alone the sregs of this instruction, for
+	// deadce to kill the instructions initializing them.
 	mono_interp_stats.constant_folds++;
-	if (sp [0].ins != NULL && sp [1].ins != NULL) {
-		interp_clear_ins (sp [0].ins);
-		interp_clear_ins (sp [1].ins);
-		mono_interp_stats.killed_instructions += 2;
-		if (result.type == STACK_VALUE_I4)
-			ins = interp_get_ldc_i4_from_const (td, ins, result.i);
-		else if (result.type == STACK_VALUE_I8)
-			ins = interp_inst_replace_with_i8_const (td, ins, result.l);
-		else
-			g_assert_not_reached ();
-		if (td->verbose_level) {
-			g_print ("Fold binop :\n\t");
-			dump_interp_inst (ins);
-		}
-		sp [0].ins = ins;
-	} else {
-		sp [0].ins = NULL;
+
+	if (result.type == LOCAL_VALUE_I4)
+		ins = interp_get_ldc_i4_from_const (td, ins, result.i, dreg);
+	else if (result.type == LOCAL_VALUE_I8)
+		ins = interp_inst_replace_with_i8_const (td, ins, result.l);
+	else
+		g_assert_not_reached ();
+
+	if (td->verbose_level) {
+		g_print ("Fold binop :\n\t");
+		dump_interp_inst (ins);
 	}
-	sp [0].val = result;
-	return ins;
 
-cfold_failed:
-	sp->ins = NULL;
-	sp->val.type = STACK_VALUE_NONE;
+	local_ref_count [sreg1]--;
+	local_ref_count [sreg2]--;
+	local_defs [dreg] = result;
 	return ins;
 }
 
-#define INTERP_FOLD_BINOP_BR(_opcode,_stack_type,_cond) \
+// Due to poor current design, the branch op might not be the last instruction in the bblock
+// (in case we fallthrough and need to have the stack locals match the ones from next_bb, done
+// in fixup_newbb_stack_locals). If that's the case, clear all these mov's. This helps bblock
+// merging quickly find the MINT_BR_S opcode.
+#define INTERP_FOLD_BINOP_BR(_opcode,_local_type,_cond) \
 	case _opcode: \
-		g_assert (sp [0].val.type == _stack_type); \
-		g_assert (sp [1].val.type == _stack_type); \
+		g_assert (val1->type == _local_type); \
+		g_assert (val2->type == _local_type); \
 		if (_cond) { \
 			ins->opcode = MINT_BR_S; \
 			if (cbb->next_bb != ins->info.target_bb) \
 				interp_unlink_bblocks (cbb, cbb->next_bb); \
+			for (InterpInst *it = ins->next; it != NULL; it = it->next) \
+				interp_clear_ins (it); \
```

Comment:
If you delete the movs, don't you need to run fixup_newbb_stack_locals again after cprop if basic block merging can't merge these blocks? Can't that happen if the target block has an incount > 1?

Reply:
I don't fully understand the question. I think it is not clear what is happening here. Take this example :

BB1:
     mov 1 <- something (this is on the execution stack, accessed both from BB2 and BB3)
     mov 2 <- true
     brtrue 2 BB3
     mov 3 <- 1 (this mov is added by the fixup method)
BB2:
    do something with vreg 3
BB3:
    do something with vreg 1

BB2 and BB3 both expect something on the IL stack, however they reference it through different vregs. brtrue is resolved to a direct branch to BB3. The moves in this case are only for the followthrough case, and they become unreachable so they can be removed. There is no real change in the code that would require additional moves.